UI Tests for username change on /settings page.

petercr opened the pull request:
# ✨ Codu Pull Request 💻

Fixes #1168 
** Only Part 3.**

## Pull Request details

Added UI tests for `/settings` to:
- tests for input field & types
- tries to submit a bad user name
- checks for error visibility
- resets the form
- changes user name, saves, and checks for update

## Any Breaking changes
None

## Associated Screenshots
None, but the tests pass 😅 I promise!

## [Optional] What gif best describes this PR or how it makes you feel

Like this, only instead of Monday, Wed, Friday it's like start, middle, end 😆 
![A dog that starts out grumpy, then get glad, then gets happy.](https://media3.giphy.com/media/v1.Y2lkPTc5MGI3NjExNnJuNXZwYXNleXhsMWV6YThxNjkyaTFwZHk2OXJjNGZybTF2MXg4OCZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/yxP4xfvgpR9jAgGrWV/giphy.gif)

petercr committed:
feat: UI tests for username change on /settings

petercr committed:
Merge branch 'codu-code:develop' into develop

JohnAllenTech commented on e2e/settings.spec.ts:
```diff
@@ -1,4 +1,4 @@
+import { test, expect } from "@playwright/test";
 import { loggedInAsUserOne } from "./utils";
 
 test.describe("Unauthenticated setttings Page", () => {
```
Can we change this to authenticated settings page. Because the user is authenticated for these tests

petercr replied:
I'm not sure exactly what you're saying here? 

Did you want a separate file for auth vs. un-auth tests, because doesn't a user have to be logged in to edit their profile?

I guess I could use more detail on this 👌🏻

JohnAllenTech replied:
Your new test is ran inside a test suite called 'Unauthenticated setttings Page'

But this is an authenticated test because users need to login to change their username.

If you just change the name of the test it will be fixed

JohnAllenTech commented on e2e/settings.spec.ts:
```diff
@@ -10,6 +10,32 @@ test.describe("Authenticated settings Page", () => {
+    // Test that the input field can be filled with a valid value and saves it
+    await inputField.fill('codu-rules');
+    await page.locator('button[type="submit"]').click(); 
+    await expect(inputField).toHaveValue('codu-rules');
```
Will this wait for the username change to take effect? 

If you fill in the new username and then click submit.

Will this test see the new username you have entered as the value of the input and complete straight away?

petercr replied:
I mean the tests pass for it. But it could be just catching the still filled in form from before the reset. 

Would you like me to change it to a page refresh instead? 

Or would you like some kind of timeOut in-between the state change and checking for the change to persist?

I would be happy to add either one 👍🏻

JohnAllenTech replied:
Yeah I am not sure is there a loading indicator for these types of change? It would be nice to see

1. Click submit
2. Loading indicator appears
3. Loading indicator hidden
4. Check new username has taken effect

JohnAllenTech reviewed (changes requested):
Just made some general comments. I am not a maintainer here so everything is just a suggestion

NiallJoeMaher commented:
@JohnAllenTech let me know when this is good to go since you are the overlord of tests.

petercr committed:
Merge branch 'codu-code:develop' into develop

petercr committed:
fix: without adding "https://" a form error prevents /settings testing.

petercr committed:
feat: add class names to use with UI testing.

petercr committed:
feat: new tests to check for toast message and field change after reset.

petercr commented:
@JohnAllenTech I made some changes and added the extra tests you had suggested. 
Tests now cover:

- waiting for toast to show
- waiting for toast to disappear
- full page refresh
- checking new value persists in input field

I did have to change a field in `/e2e/setup/setup.ts` to add `https://` to it.
Otherwise you get an error on the form in E2E testing.

I also added new class names to both the toast messages for UI testing.

Thanks

petercr commented:
@JohnAllenTech I made some changes and added the extra tests you had suggested. 
Tests now cover:

- waiting for toast to show
- waiting for toast to disappear
- full page refresh
- checking new value persists in input field

I did have to change a field in `/e2e/setup/setup.ts` to add `https://` to it.
Otherwise you get an error on the form.

petercr commented:
I made some changes and added the extra tests you had suggested. 
Tests now cover:

- waiting for toast to show
- waiting for toast to disappear
- full page refresh
- checking new value persists in input field

I did have to change a field in `/e2e/setup/setup.ts` to add `https://` to it.
Otherwise you get a error on the form.

petercr commented on e2e/settings.spec.ts:
```diff
@@ -10,6 +10,38 @@ test.describe("Authenticated settings Page", () => {
+    // Test that the input field can be filled with a valid value and saves it
+    await inputField.fill('codu-rules');
+    await page.locator('button[type="submit"]').click(); 
+    const toastError = page.locator('.toast-success');
+    await expect(toastError).toBeVisible();
+    await expect(toastError).toBeHidden();
```
I am not sure there is even a loading spinner? 🤷🏻‍♂️

I can add the tests for user name length and the corresponding errors if you want me to.

I guess I will just add them anyways 👌🏻

petercr replied:
I have added the extra tests for min and max length on the user name input.
This should resolve this issue.

petercr committed:
fix: added tests for min and max username values and error messages.

JohnAllenTech commented:
> @JohnAllenTech I made some changes and added the extra tests you had suggested. 
> 
> Tests now cover:
> 
> 
> 
> - waiting for toast to show
> 
> - waiting for toast to disappear
> 
> - full page refresh
> 
> - checking new value persists in input field
> 
> 
> 
> I did have to change a field in `/e2e/setup/setup.ts` to add `https://` to it.
> 
> Otherwise you get an error on the form in E2E testing.
> 
> 
> 
> I also added new class names to both the toast messages for UI testing.
> 
> 
> 
> Thanks 

Hey @petercr apologies I was AFK for a few days. I'll take a look at this in a while and reply properly

CarolinaCobo commented on e2e/settings.spec.ts:
```diff
@@ -10,6 +10,32 @@
+    // Test that the input field can be filled with a valid value and saves it
+    await inputField.fill('codu-rules');
+    await page.locator('button[type="submit"]').click(); 
+    await expect(inputField).toHaveValue('codu-rules');
```
@petercr this seems like a good addition as well if you have the time :)

NiallJoeMaher committed:
Merge branch 'develop' into develop

NiallJoeMaher commented:
Tests failing here.

petercr commented:
Well crap 😕 that's no good. 
I will look into this over the weekend (next 2 days) and see if I can figure it out. 👌🏻

I wonder if this has to do with the newsletter sub function?
Either way I'm on it 👍🏻

petercr committed:
Merge branch 'develop' into develop

petercr commented:
When I run the test locally, they all pass 🤷🏼‍♂️ see pic below. 
I'm not sure but this may have to do with `/server/lib/newsletter.ts`.

I added extra code on line 39 to keep from getting the email API errors for the newsletter. 
You can see some controller errors from `posts.sidebarData` and  the tRPC error for `profile.edit`.

Here's what I added locally for this:

```js
 if (process.env.ENV === "local" || "development") {
    return true;
  }
```

But it might be simpler to go with:

```js
 if (process.env.ENV !== "production") {
    return true;
  }
```

Here is the image of the tests passing.

![msrdc_ttyjHMMI1j](https://github.com/user-attachments/assets/a58aba08-b46d-47d6-afc5-b87bef00942b)
